taggit/admin.py

```diff
@@ -14,3 +19,66 @@ class TagAdmin(admin.ModelAdmin):
     ordering = ["name", "slug"]
     search_fields = ["name"]
     prepopulated_fields = {"slug": ["name"]}
+    actions = ["render_tag_form"]
+
+    def get_urls(self):
+        urls = super().get_urls()
+        custom_urls = [
+            path(
+                "merge-tags/",
+                self.admin_site.admin_view(self.merge_tags_view),
+                name="merge_tags",
+            ),
+        ]
+        return custom_urls + urls
+
+    @admin.action(description="Merge selected tags")
+    def render_tag_form(self, request, queryset):
+        selected = request.POST.getlist(admin.helpers.ACTION_CHECKBOX_NAME)
+        if not selected:
+            self.message_user(request, "Please select at least one tag.")
+            return redirect(request.get_full_path())
+
+        selected_tag_ids = ",".join(selected)
+        redirect_url = (
+            f"{request.get_full_path()}merge-tags/?selected_tags={selected_tag_ids}"
+        )
+
+        return redirect(redirect_url)
+
+    def merge_tags_view(self, request):
+        if request.method == "GET":
+            selected_tag_ids = request.GET.get("selected_tags", "").split(",")
+
+            # store selected_tag_ids in session data until they are merged
+            request.session["selected_tag_ids"] = selected_tag_ids
+        else:
+            selected_tag_ids = request.session.get("selected_tag_ids", [])
+
+        if request.method == "POST":
+            form = MergeTagsForm(request.POST)
+            if form.is_valid():
+                new_tag_name = form.cleaned_data["new_tag_name"]
+                new_tag, created = Tag.objects.get_or_create(name=new_tag_name)
+                with transaction.atomic():
+                    for tag_id in selected_tag_ids:
+                        tag = Tag.objects.get(id=tag_id)
+                        tagged_items = TaggedItem.objects.filter(tag=tag)
+                        for tagged_item in tagged_items:
+                            tagged_item.tag = new_tag
+                            tagged_item.save()
+                        # tag.delete()  #this will delete the selected tags after merge
+
+                # clear the selected_tag_ids from session after merge is complete
+                request.session.pop("selected_tag_ids", None)
+                return redirect("..")
```

Comment:
Does that work? TIL

If that doesn't work and you want to always redirect to the admin list view, there is some name given by the admin (I think it's `admin_model_list_view`? With model names as params).

Reply:
Also, I think it would be worth calling `self.message_user` with a success message when we successfully "do the thing", so to speak.

Reply:
The redirect does work. Since I am appending the `merge-tags/` during the first redirect to the form,  this goes back one level to the path before which is the list of tags.

Reply:
I added the `self.message_user` to notify the user after the merge was complete but I think the redirect back to the previous page is happen too quickly to see it. Maybe we can find a better way to notify the users.

Reply:
Hmmm... normally the mesasge should appear on the "next" page load so to speak (if this is happening through the Django `messages` framework), will take a look.

Reply:
Try `self.message_user(request, "some message", messages.SUCCESS)`, that should work (I looked inside the definition of `ModelAdmin` to find some examples)

Reply:
the `messages.SUCCESS` didn't work here but `level="success` got it working